# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [63]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
os.listdir()

['.DS_Store',
 'GuidedCapstoneStep4 and step5HL.ipynb',
 'Step2',
 'GuidedCapstoneStep3HL.ipynb',
 'models',
 'updated_ski_data_.csv',
 'updated_ski_data_.numbers',
 'Submitted',
 'figures',
 '.ipynb_checkpoints',
 'data']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [99]:
file="/Users/omniverse/Documents/Springboard/Capstone/redone/data/step3_output.csv"
df=pd.read_csv(file)
df.head().T

0                    1  \
Name               Hilltop Ski Area  Sunrise Park Resort   
state                        Alaska              Arizona   
summit_elev                    2090                11100   
vertical_drop                   294                 1800   
trams                             0                    0   
fastEight                         0                    0   
fastSixes                         0                    0   
fastQuads                         0                    1   
quad                              0                    2   
triple                            1                    3   
double                            0                    1   
surface                           2                    0   
total_chairs                      3                    7   
Runs                             13                   65   
TerrainParks                      1                    2   
LongestRun_mi                     1                  1.2   
SkiableTerrain_ac                30                  800   
Snow Making_ac                   30                   80   
daysOpenLastYear                150                  115   
yearsOpen                        36                   49   
averageSnowfall                  69                  250   
AdultWeekday                     30                   74   
AdultWeekend                     34                   78   
projectedDaysOpen               152                  104   
NightSkiing_ac                   30                   80   
clusters                          0                    1   

                                               2                       3  \
Name               Yosemite Ski & Snowboard Area  Boreal Mountain Resort   
state                                 California              California   
summit_elev                                 7800                    7700   
vertical_drop                                600                     500   
trams                                          0                       0   
fastEight                                      0                       0   
fastSixes                                      0                       0   
fastQuads                                      0                       1   
quad                                           0                       1   
triple                                         1                       3   
double                                         3                       1   
surface                                        1                       2   
total_chairs                                   5                       8   
Runs                                          10                      33   
TerrainParks                                   2                       6   
LongestRun_mi                                0.4                       1   
SkiableTerrain_ac                             88                     380   
Snow Making_ac                                 0                     200   
daysOpenLastYear                             110                     150   
yearsOpen                                     84                      54   
averageSnowfall                              300                     400   
AdultWeekday                                  47                      49   
AdultWeekend                                  47                 64.1668   
projectedDaysOpen                            107                     150   
NightSkiing_ac                                 0                     200   
clusters                                       1                       1   

                             4  
Name               Dodge Ridge  
state               California  
summit_elev               8200  
vertical_drop             1600  
trams                        0  
fastEight                    0  
fastSixes                    0  
fastQuads                    0  
quad                         1  
triple                       2

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [67]:
df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)
print(dfo.shape)
dfo.head()

(176, 60)


Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [68]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [69]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [70]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [71]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [72]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [73]:
explained_variance_score(y_test, y_pred)

-1.4286310549339054e+19

In [74]:
mean_absolute_error(y_test, y_pred)

7861421563.285052

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [75]:
lm.intercept_

1965355445.5558972

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [76]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeffdf = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

coeffdf.head(10)

Coefficient
summit_elev    7.687717e-02
vertical_drop  3.087621e+00
trams          2.483742e+11
fastEight      1.670232e+12
fastSixes      7.524255e+11
fastQuads      1.472857e+11
quad           1.762210e+11
triple         3.094828e+11
double         3.788897e+11
surface        3.625902e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [90]:
file="/Users/omniverse/Documents/Springboard/Capstone/redone/data/step3_output.csv"
df=pd.read_csv(file)
df.head()

#df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)
print(df.shape)
df.head()

(176, 26)


Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0            80.0        115.000000       49.0   
2               88.0             0.0        110.000000       84.0   
3              380.0           200.0        150.000000       54.0   
4              862.0             0.0        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

In [91]:
X=df.drop(['AdultWeekend','Name','state'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)


In [92]:
explained_variance_score(y_test, y_pred)

0.5199029021516772

In [93]:
mean_absolute_error(y_test, y_pred)

7.534344353225304

In [94]:
coeffdf = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

coeffdf.head(10)

Coefficient
summit_elev    3.039686e+00
vertical_drop  2.770467e+00
trams          2.220446e-15
fastEight      1.776357e-15
fastSixes      4.440892e-16
fastQuads      3.966489e-01
quad           1.971225e+00
triple         3.211720e-01
double         3.592714e-01
surface        1.156217e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [95]:
file="/Users/omniverse/Documents/Springboard/Capstone/redone/data/step3_output.csv"
df=pd.read_csv(file)
df.head()

#df = pd.concat([df, pd.get_dummies(df['state'])], axis=1).drop(['state'], axis =1)
print(df.shape)
df.head()

(176, 26)


Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0            80.0        115.000000       49.0   
2               88.0             0.0        110.000000       84.0   
3              380.0           200.0        150.000000       54.0   
4              862.0             0.0        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 26 columns]

In [100]:
#'base_elev' column is missing.
X=df.drop(['AdultWeekend','Name','state','summit_elev'], axis=1)
y=df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
y=y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)



In [101]:
explained_variance_score(y_test, y_pred)

0.5234888253827603

In [102]:
mean_absolute_error(y_test, y_pred)

7.516552245262175

In [103]:
coeffdf = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

coeffdf.head(10)

Coefficient
vertical_drop  2.747260e+00
trams          0.000000e+00
fastEight      1.776357e-15
fastSixes      8.881784e-16
fastQuads      5.254315e-01
quad           1.890437e+00
triple         6.916117e-02
double         4.053185e-01
surface        8.079475e-01
total_chairs   4.238774e-01

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.428 | 7861421563.285 |-|
| Model 2. | 0.52| 7.534 |'state'|
| Model 3. | 0.523 | 7.57 |'state','summit_elev','base_elev'|

Model Selection: